File: go/_2021/day1.go
package _2021

import (
	"fmt"
	"strconv"
	"strings"

	"advent/utils"
)

func parseInput() []int {
	data := utils.LoadPuzzleInput("2021/day1")
	input_strings := strings.Fields(data)
	integers := make([]int, len(input_strings))
	for ind, str := range input_strings {
		i, _ := strconv.Atoi(str)
		integers[ind] = i
	}
	return integers
}

func Day1Part1() string {
	integers := parseInput()
	increases := 0
	previousDepth := 0
	for ii, depth := range integers {
		if ii == 0 {
			continue
		}
		if depth > previousDepth {
			increases++
		}
		previousDepth = depth
	}
	return fmt.Sprint(increases)
}

func Day1Part2() string {
	integers := parseInput()
	increases := 0
	windowWidth := 3
	previous := integers[0] + integers[1] + integers[2]
	for ii := windowWidth + 1; ii < len(integers)+1; ii++ {
		sum := 0
		for _, value := range integers[ii-windowWidth : ii] {
			sum += value
		}
		if sum > previous {
			increases++
		}
		previous = sum
	}
	return fmt.Sprint(increases)
}

func Day1() {
	utils.Profile(Day1Part1)
	utils.Profile(Day1Part2)
}

File: python/_2021/day1.py
from python import utils

raw = utils.load_puzzle_input("2021/day1")

input = list(map(int, map(str.strip, raw.strip().split("\n"))))


@utils.profile
def part1():
    increases = 0
    previous_depth = 0
    for ii, depth in enumerate(input[1:]):
        if depth > previous_depth:
            increases += 1
        previous_depth = depth
    return increases


@utils.profile
def part2():
    increases = 0
    for ii, depth in enumerate(input):
        current_window = input[ii : ii + 3]
        next_window = input[ii + 1 : ii + 4]
        if len(next_window) < 3:
            break  # reached end of list
        if sum(next_window) > sum(current_window):
            increases += 1

    return increases


if __name__ == "__main__":
    assert part1() == 1482
    assert part2() == 1518
